src/DmaInputFilter.cc

```diff
@@ -0,0 +1,132 @@
+#include <cassert>
+#include <cstdio>
+#include <cerrno>
+#include <system_error>
+#include <iostream>
+#include <sstream>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include "DmaInputFilter.h"
+#include "log.h"
+
+
+
+DmaInputFilter::DmaInputFilter( const std::string& deviceFileName, size_t packetBufferSize, size_t nbPacketBuffers, ctrl& control ) : 
+  InputFilter( packetBufferSize, nbPacketBuffers, control )
+{ 
+  dma_fd = open( deviceFileName.c_str(), O_RDWR | O_NONBLOCK );
+  if ( dma_fd < 0 ) {
+    throw std::system_error(errno, std::system_category(), "Cannot open DMA device: " + deviceFileName);
+  }
+
+  LOG(TRACE) << "Created DMA input filter"; 
+}
+
+DmaInputFilter::~DmaInputFilter() {
+  close( dma_fd );
+  LOG(TRACE) << "Destroyed DMA input filter";
+}
+
+/*
+ * man 2 write:
+ * On Linux, write() (and similar system calls) will transfer at most
+ * 	0x7ffff000 (2,147,479,552) bytes, returning the number of bytes
+ *	actually transferred.  (This is true on both 32-bit and 64-bit
+ *	systems.)
+ */
+
+#define RW_MAX_SIZE	0x7ffff000
+
+static inline ssize_t read_axi_packet_to_buffer(int fd, char *buffer, uint64_t size)
+{
+  ssize_t rc;
+  uint64_t to_read = size;
+
+  if (to_read > RW_MAX_SIZE) {
+    to_read = RW_MAX_SIZE;
+  }
+
+  /* read data from file into memory buffer */
+  rc = read(fd, buffer, to_read);
+
+  if (rc <= 0) {
+    return rc;
+  }	
+  return rc;
+}
+
+ssize_t DmaInputFilter::readPacketFromDMA(char **buffer, size_t bufferSize)
+{
+  // Read from DMA
+  // We need at least 1MB buffer
+
+  ssize_t bytesRead = 0;
+  int skip = 0;
+
+  while (true) {
+    // Read from DMA
+    bytesRead = read_axi_packet_to_buffer(dma_fd, *buffer, bufferSize);
+
+    if (bytesRead < 0) {
+      skip++;
+      // Check for errors and then skip
+      if (errno == EIO || errno == EMSGSIZE) {
+        if (errno == EIO) {
+          stats.nbDmaErrors++;
+          LOG(ERROR) << "#" << nbReads() << ": DMA I/O ERROR. Skipping packet #" << skip << '.';
+        } else {
+          stats.nbDmaOversizedPackets++;
+          LOG(ERROR) << "#" << nbReads() << ": DMA read returned oversized packet. DMA returned " << bytesRead << ", buffer size is " << bufferSize << ". Skipping packet #" << skip << '.';
+        }
+        continue;
+      }
+
+      // Some fatal error occurred
+      std::ostringstream os;
+      os  << "Iteration: " << nbReads() << "  ERROR: DMA read failed.";
+      throw std::system_error(errno, std::system_category(), os.str() );
+    }
+
+    // We have some data
+    break;
+  }
+
+  // This should not happen
+  if (bytesRead > (ssize_t)bufferSize ){
+    std::ostringstream os;
+    os  << "Iteration: " << nbReads()
+        << "  ERROR: DMA read returned " << bytesRead 
+        << " > buffer size " << bufferSize;
+    throw std::runtime_error( os.str() );
```

Comment:
I didn't follow this to right to the top of the hierarchy, but do we catch this somewhere? Otherwise maybe stating in the message that this indicates a problem in the firmware might be useful for people going through the logs when something's wrong at P5.

Reply:
It is not catch, because the scdaq cannot recover from this condition. So it would end with a stack trace.
 
In fact, in this DMA input filer it cannot happen, so I will remove it. Still, I will have a look why we did have it here in the past.

Reply:
Removed.